### PR TITLE
[core] Fix cpp inclusion check

### PR DESCRIPTION
**Pull request by dentiny**
To resolve issue https://github.com/ray-project/ray/pull/49295#issuecomment-2547840327

### CODE REVIEW

File: ci/lint/check_cpp_files_inclusion.py [hunk 1]
```diff
@@ -3,9 +3,10 @@
 """
 
 import subprocess
+import re
 
-# All ray core C++ code is under certain folder.
-_RAY_CORE_CPP_DIR = "src/ray"
+# Regex to filter ray core C++ filename.
+_RAY_CORE_CPP_REGEX = r"^src/ray.*\.(h|cc)$"
```
Comment (dentiny) [inline ci/lint/check_cpp_files_inclusion.py:9]: we should only check C++ files, which is with suffix `.h` or `.cc`

**Comment by MortalHappiness** [comment]
You should change `.pre-commit-config.yaml` to something like this:

```yaml
  - repo: local
    hooks:
      - id: check_cpp_files_inclusion
        name: Check ray core C++ files inclusion violations
        entry: python ci/lint/check_cpp_files_inclusion.py
        language: python
        files: '^src/ray/'
        types: [c++]
```

Because we check for c++ files, not python files.

And then in `ci/lint/check_cpp_files_inclusion.py`, use `sys.argv` to get the filename that `pre-commit` passed to you.

When `pre-commit` calls `entry`, `sys.argv` will look like this:

`['ci/lint/check_cpp_files_inclusion.py', 'src/ray/raylet/scheduling/local_resource_manager.h', 'src/ray/core_worker/lib/java/io_ray_runtime_util_SystemConfig.h', 'src/ray/gcs/gcs_server/gcs_server_io_context_policy.h', 'src/ray/rpc/worker/core_worker_client.h', 'src/ray/object_manager/plasma/malloc.cc', 'src/ray/object_manager/ownership_based_object_directory.cc', 'src/ray/gcs/gcs_client/test/gcs_client_reconnection_test.cc', 'src/ray/rpc/server_call.cc']`

`pre-commit` may call `entry` multiple times.

And in `ci/lint/check_cpp_files_inclusion.py`, don't rely on the external `grep` command, use the Python `re` library instead.

**Comment by dentiny** [comment]
> language: python
files: '^src/ray/'
types: [c++]

I'm not sure it works, for example, it doesn't produce files that we changed.

**Comment by MortalHappiness** [comment]
> I'm not sure it works, for example, it doesn't produce files that we changed.

I am pretty sure it produces files that we changed. Maybe you can upgrade `pre-commit` to the latest version and try again?

**Comment by dentiny** [comment]
> I am pretty sure it produces files that we changed. Maybe you can upgrade `pre-commit` to the latest version and try again?

For the files you listed above (9 files in total), did you change all of them?
Are they in staging area (after `git add` but not `git commit`), or already committed?
Did you run the command on your branch or master branch?

**Comment by dentiny** [comment]
Synced offline with @MortalHappiness , thanks for the help! Using `re` and pre-commit hook feature is much better!

File: ci/lint/check_cpp_files_inclusion.py [hunk 2]
```diff
@@ -38,18 +14,19 @@ def check_ray_core_inclusion(fname: str):
...
-    cmd = f"grep 'src/ray' {fname} | grep -v 'pb'"
-    grep_res = subprocess.check_output(cmd, shell=True, text=True)
-    if grep_res:
-        raise ImportError(f"{fname} has invalid header file inclusion: {grep_res}")
+    pattern = re.compile(r"^(?=.*src/ray)(?!.*pb)")
+
+    with open(fname, "r") as file:
+        for line in file:
+            if pattern.search(line):
+                raise ImportError(f"{fname} has invalid header file inclusion: {line}")
```
Comment (MortalHappiness) [inline ci/lint/check_cpp_files_inclusion.py:22]: I suggest you to raise user-defined error here. [ImportError](https://docs.python.org/3.12/library/exceptions.html#ImportError) is used when the `import` statement has troubles, so raise `ImportError` here is confusing.

You can define a custom error like this:

```python
class InvalidCppHeaderInclusion(Exception):
    pass
```

And then `raise InvalidCppHeaderInclusion`

**Review by MortalHappiness — CHANGES_REQUESTED** [review]

File: ci/lint/check_cpp_files_inclusion.py [hunk 3]
```diff
@@ -38,18 +14,19 @@ def check_ray_core_inclusion(fname: str):
...
 
 
 def main():
-    changed_files = diff_files_with_develop()
+    changed_files = sys.argv[1:]
     for cur_file in changed_files:
-        if not cur_file.startswith(_RAY_CORE_CPP_DIR):
-            continue
         check_ray_core_inclusion(cur_file)
+    exit(0)
```
Comment (MortalHappiness) [inline ci/lint/check_cpp_files_inclusion.py:29]: This line can be removed.

File: ci/lint/check_cpp_files_inclusion.py (hunk 2 above)
Comment (MortalHappiness) [inline ci/lint/check_cpp_files_inclusion.py:22]: Use `file.read()` to read the whole file, and then use multiline mode for `re`. See https://docs.python.org/3/library/re.html#re.M for details.

File: ci/lint/check_cpp_files_inclusion.py (hunk 3 above)
Comment (dentiny) [reply ci/lint/check_cpp_files_inclusion.py:32]: Updated.

File: ci/lint/check_cpp_files_inclusion.py (hunk 2 above)
Comment (dentiny) [reply ci/lint/check_cpp_files_inclusion.py]: Updated.

File: ci/lint/check_cpp_files_inclusion.py [hunk 4]
```diff
@@ -2,53 +2,31 @@
...
-    if grep_res:
-        raise ImportError(f"{fname} has invalid header file inclusion: {grep_res}")
+    pattern = re.compile(r"^(?=.*src/ray)(?!.*pb).*$", re.MULTILINE)
+
+    with open(fname, "r") as file:
+        content = file.read()
+        match = pattern.search(content)
+        if match:
+            print(f"{fname} has invalid header file inclusion: {match.group(0)}")
+            exit(1)
```
Comment (MortalHappiness) [inline ci/lint/check_cpp_files_inclusion.py:24]: nit: https://stackoverflow.com/questions/6501121/difference-between-exit-and-sys-exit-in-python

```suggestion
            sys.exit(1)
```

**Review by MortalHappiness — APPROVED** [review]

**Comment by MortalHappiness** [comment]
Do you want to fix these errors in this PR or a separate PR?

image

File: ci/lint/check_cpp_files_inclusion.py (hunk 4 above)
Comment (dentiny) [reply ci/lint/check_cpp_files_inclusion.py:24]: Lesson learnt, updated.

**Comment by dentiny** [comment]
> Do you want to fix these errors in this PR or a separate PR?

Thank you! Updated the regex to handle such case.

**Comment by MortalHappiness** [comment]
It seems that there are still errors

image

**Pull request merged**